### PR TITLE
feat: add details to rest exception

### PR DESCRIPTION
**Pull request by ashish-s**
# Feature #

Add a new property named `details` to RestException. Details is returned by selective apis to populate extra information about errors.
The Studio V2 flow api will be returning extra details with flow validation errors.

### Checklist
- [x] I acknowledge that all my contributions will be made under the project's license
- [x] I have made a material change to the repo (functionality, testing, spelling, grammar)
- [x] I have read the [Contribution Guidelines](CONTRIBUTING.md) and my PR follows them
- [x] I have titled the PR appropriately
- [x] I have updated my branch with the master branch
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation about the functionality in the appropriate .md file
- [x] I have added inline documentation to the code I modified

If you have questions, please file a [support ticket](https://twilio.com/help/contact), or create a GitHub Issue in this repository.

### CODE REVIEW

**Comment by ashish-s** [comment]
> LGTM, thank you!
> 
> I think we just need a test(s).

I tried to find existing tests for RestExceptions, but didnt see any, is there a reference test i can follow ?

**Comment by thinkingserious** [comment]
I would go with a new test in the [`Rest`](https://github.com/twilio/twilio-php/tree/master/tests/Twilio/Unit/Rest) folder, similar to what we have in [`ClientTest.php`](https://github.com/twilio/twilio-php/blob/master/tests/Twilio/Unit/Rest/ClientTest.php). Thanks!

**Comment by ashish-s** [comment]
@thinkingserious added a unit test and another property named `moreInfo`, to make it all compatible with other libs.

File: src/Twilio/Exceptions/RestException.php [hunk 1]
```diff
@@ -6,17 +6,23 @@
...
      * @param int $statusCode [optional] The HTTP Status code.
+     * @param string $moreInfo [optional] More information about the error.
+     * @param array $details [optional] Additional details about the error.
      * @since 5.1.0
      */
-    public function __construct(string $message, int $code, int $statusCode) {
+    public function __construct(string $message, int $code, int $statusCode, string $moreInfo = null, array $details = null) {
         $this->statusCode = $statusCode;
+        $this->details = $details;
+        $this->moreInfo = $moreInfo;
```
Comment (childish-sambino) [inline src/Twilio/Exceptions/RestException.php:25]: nit: ordering differs from param order.

File: src/Twilio/Page.php [hunk 2]
```diff
@@ -45,13 +45,17 @@ protected function processResponse(Response $response) {
...
 
             $content = $response->getContent();
+            $details = null;
+            $moreInfo = null;
 
             if (\is_array($content)) {
                 $message .= isset($content['message']) ? ': ' . $content['message'] : '';
                 $code = $content['code'] ?? $code;
+                $details = isset($content['details']) ? $content['details'] : null;
+                $moreInfo = isset($content['more_info']) ? $content['more_info'] : null;
```
Comment (childish-sambino) [inline src/Twilio/Page.php:55]: Same nits.

File: src/Twilio/Page.php [hunk 3]
```diff
@@ -45,13 +45,17 @@ protected function processResponse(Response $response) {
             $code = $response->getStatusCode();
 
             $content = $response->getContent();
+            $details = null;
+            $moreInfo = null;
 
             if (\is_array($content)) {
                 $message .= isset($content['message']) ? ': ' . $content['message'] : '';
                 $code = $content['code'] ?? $code;
+                $details = isset($content['details']) ? $content['details'] : null;
```
Comment (childish-sambino) [inline src/Twilio/Page.php:54]: Could this just be `$details = $content['details'];`?

File: src/Twilio/Version.php [hunk 4]
```diff
@@ -83,7 +83,9 @@ protected function exception(Response $response, string $header): TwilioExceptio
         if (\is_array($content)) {
             $message .= isset($content['message']) ? ': ' . $content['message'] : '';
             $code = isset($content['code']) ? $content['code'] : $response->getStatusCode();
-            return new RestException($message, $code, $response->getStatusCode());
+            $details = isset($content['details']) ? $content['details'] : $response->getDetails();
```
Comment (childish-sambino) [inline src/Twilio/Version.php:86]: The response object doesn't have a `getDetails` method.

File: tests/Twilio/Unit/VersionTest.php [hunk 5]
```diff
@@ -223,6 +228,30 @@ public function testAbsoluteUrl(string $message, string $prefix, string $uri, st
...
+                                    }'));
+        try {
+            $this->version->fetch('get', 'http://foo.bar');
+            self::fail();
+        }catch (RestException $rex){
+            $this->assertEquals(20001, $rex->getCode());
+            $this->assertEquals(400, $rex->getStatusCode());
+            $this->assertEquals('[HTTP 400] Unable to fetch record: Bad request', $rex->getMessage());
+            $this->assertEquals('https://www.twilio.com/docs/errors/20001', $rex->getMoreInfo());
+            $this->assertEquals(array("foo" => "bar"), $rex->getDetails());
```
Comment (childish-sambino) [inline tests/Twilio/Unit/VersionTest.php:251]: nit: prefer `[]` over `array()`

File: src/Twilio/Exceptions/RestException.php (hunk 1 above)
Comment (ashish-s) [reply src/Twilio/Exceptions/RestException.php:24]: fixed

File: src/Twilio/Page.php (hunk 2 above)
Comment (ashish-s) [reply src/Twilio/Page.php]: fixed

File: src/Twilio/Page.php (hunk 3 above)
Comment (ashish-s) [reply src/Twilio/Page.php]: i tried doing that, but it complains with `no index error`

File: src/Twilio/Version.php (hunk 4 above)
Comment (ashish-s) [reply src/Twilio/Version.php]: removed

File: tests/Twilio/Unit/VersionTest.php (hunk 5 above)
Comment (ashish-s) [reply tests/Twilio/Unit/VersionTest.php]: fixed

**Comment by ashish-s** [comment]
@childish-sambino fixed formatting issues, added few more tests to make all paths get exercised

File: src/Twilio/Page.php (hunk 3 above)
Comment (childish-sambino) [reply src/Twilio/Page.php]: Ah, yes. Well this syntax is less verbose: `$details = $content['details'] ?? null;`

Comment (ashish-s) [reply src/Twilio/Page.php]: that works ! fixed

**Pull request merged**